src/backend/distributed/operations/auto_shard_split.c

```diff
@@ -0,0 +1,370 @@
+#include "postgres.h"
+#include "libpq-fe.h"
+#include "executor/spi.h"
+#include "distributed/lock_graph.h"
+#include "distributed/coordinator_protocol.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/metadata_utility.h"
+#include "distributed/multi_logical_replication.h"
+#include "postmaster/postmaster.h"
+#include "distributed/distribution_column.h"
+#include "utils/builtins.h"
+#include "distributed/shard_split.h"
+
+
+PG_FUNCTION_INFO_V1(citus_auto_shard_split_start);
+
+int MaxShardSize = 102400;
+
+/*
+ * Struct to store all the information related to
+ * a shard.
+ */
+typedef struct ShardInfoData
+{
+	int64 shardsize;
+	int64 shardminvalue;
+	int64 shardmaxvalue;
+	int64 shardid;
+	int32 nodeid;
+	char *tablename;
+	char *distributionColumn;
+	char *datatype;
+	char *shardname;
+	Oid tableId;
+	Oid distributionColumnId;
+}ShardInfoData;
+typedef ShardInfoData *ShardInfo;
+
+
+/*
+ * It throws an error if a concurrent automatic shard split or Rebalance operation is happening.
+ */
+void
+ErrorOnConcurrentOperation()
+{
+	int64 jobId = 0;
+	if (HasNonTerminalJobOfType("rebalance", &jobId))
+	{
+		ereport(ERROR, (
+					errmsg("A rebalance is already running as job %ld", jobId),
+					errdetail("A rebalance was already scheduled as background job"),
+					errhint("To monitor progress, run: SELECT * FROM "
+							"citus_rebalance_status();")));
+	}
+	if (HasNonTerminalJobOfType("Automatic Shard Split", &jobId))
```

Comment:
Yes ig we can add this since running this function is going to generate shards on a particular node which is going to be moved by the rebalancer only so both shouldn't happen concurrently